### PR TITLE
add `--in-prefix-bos` to prefix BOS to user inputs; keep EOS

### PR DESCRIPTION
**Pull request by jxy**
The BOS precedes the string specified by `--in-prefix`. Model generated EOS is now kept in the context.

It provides a way to strictly following the prompt format used in Llama-2-chat.

The EOS handling also benefits some existing finetunes that uses EOS to mark the end of turn.

### CODE REVIEW

**Comment by jxy** [comment]
For llama-2-chat, you want

```bash
$ ./main -m "$MODEL" -c 4096 -n -1 \
--in-prefix-bos --in-prefix ' [INST] ' --in-suffix ' [/INST]' -i -p \
"[INST] <<SYS>>
$SYSTEM
<</SYS>>

$instruct [/INST]"
```
The spaces in in-prefix/suffix are important.

**Comment by bullno1** [comment]
Currently prompt already has an implicit BOS right?

**Comment by ggerganov** [comment]
@bullno1 

`llama_tokenize()` has a `bool` flag whether to add a BOS or not. In `main` we set the flag to true when tokenizing the prompt

**Review by ggerganov — COMMENTED** [review]
The change seems good, but maybe do some more testing by more people to be sure we didn't mess up the main loop somehow

**Comment by JackJollimore** [comment]
> For llama-2-chat, you want
> 
> ```shell
> $ ./main -m "$MODEL" -c 4096 -n -1 \
> --in-prefix-bos --in-prefix ' [INST] ' --in-suffix ' [/INST]' -i -p \
> "[INST] <<SYS>>
> $SYSTEM
> <</SYS>>
> 
> $instruct [/INST]"
> ```
> 
> The spaces in in-prefix/suffix are important.

I'm trying, but my command line does not accept the format:
Screenshot_20230721_111519

I try to clean it, but then format is changed:
Screenshot_20230721_111551


I dunno how to force a new line. Also, am I to expected replace $SYSTEM, and $instruct or leave it alone?

**Comment by jxy** [comment]
@JackJollimore you're missing the backslashes `\` at the end of the line in your shell (and somehow extra newlines???), and having extra quotation marks `"` in your prompt file.

**Comment by arch-btw** [comment]
Unfortunately it's not working for me. It starts asking itself questions and then answers them.

**Comment by arch-btw** [comment]
Unfortunately it doesn't work for me either. It starts asking itself questions and then answers them.

**Comment by JackJollimore** [comment]
> you're missing the backslashes `\` at the end of the line in your shell (and somehow extra newlines???), and having extra quotation marks `"` in your prompt file.


I overlooked the quotation after converting to .txt. Thanks for pointing that out, but I copy/pasted, so if there's new lines then the shell did that. Termux ignored commands after backslash, and I deleted them to make it work.

I don't have a clear understanding of this PR, so maybe someone skilled will test.

**Comment by arch-btw** [comment]
Just following up that it works for me now, I made a few user errors and also ran into the copy + paste problem. It's all solved now. 👍

**Comment by zacps** [comment]
Another :+1:, working well for me as well on `llama2-70b-chat` (fp16) after rebasing on master.

**Comment by lionelchg** [comment]
Works for me as well! I have rewritten a small bash script:

```shell
#!/bin/bash

# The script should be launched like ./chat.sh models/llama-2-13b-chat.ggmlv3.q4_0.bin system_prompts/translation.txt Hello

# Load system prompt
SYSTEM_PROMPT=$(cat $2)

# Execute model
./main -m $1 -c 4096 -n -1 --in-prefix-bos --in-prefix ' [INST] ' --in-suffix ' [/INST]' -i \
    -p "[INST] <<SYS>>\n$SYSTEM_PROMPT\n<</SYS>>\n\n$3 [/INST]"
```

with the following content for `system_prompts/translation.txt`:

```shell
Translate every sentence from English into French.
```

File: examples/common.h [hunk 1]
```diff
@@ -64,6 +64,7 @@ struct gpt_params {
     std::string input_prefix      = "";  // string to prefix user inputs with
     std::string input_suffix      = "";  // string to suffix user inputs with
     std::vector<std::string> antiprompt; // string upon seeing which more user input is prompted
+    bool input_prefix_bos         = false;  // if true, prefix BOS to user inputs, before input_prefix
```
[the remark this answers is not in the document]
Comment (jxy) [reply examples/common.h]: done

**Review by ggerganov — APPROVED** [review]

**Comment by ggerganov** [comment]
@lionelchg Would be a nice contribution to `examples` - feel free to PR it

**Comment by ejones** [comment]
This is great! FYI as an additional benefit, this unblocks using `--grammar` in interactive mode as an alternative for several prompt options like `--input-prefix`, `--input-suffix` and `--reverse-prompt` (except for the unnecessary newline):
```
$ ./main -m $LLAMA2_13B_Q4_0  -i  \
  --grammar 'root ::= "### RESPONSE: *" [a-z]+ (" " [a-z]+)* "* " [^\r\n]+ "\n### HUMAN: "' \
  -p "### HUMAN: Hello, how are you?
"
...

 ### HUMAN: Hello, how are you?
### RESPONSE: *thinks for a minute* I think I'm fine.
### HUMAN: 
great!
### RESPONSE: *says something else* I have this thing that annoys me.
### HUMAN: 
```

**Comment by JackJollimore** [comment]
@jxy It appears Llama2 is the only model working as expected since this commit: https://github.com/ggerganov/llama.cpp/issues/2417

Is there something I need to do to have models other than Llama2 follow intended prompt structure?

**Comment by pugzly** [comment]
Yes, I couldn't get one of my older models (wizardlm) working with the latest llama.cpp
Until I tried manually downloading repo one commit prior to this one, and everything start working as it used to be.

**Comment by jxy** [comment]
Sorry to break people's command line with reverse prompt. Previously if you specified reverse prompt, and the model generated an EOS, the EOS is replaced by a new line and the first reverse prompt was inserted. That was a bit unintuitive and overlapped with in-prefix.

Now with this PR, if the model generates EOS, the EOS is kept in the context, and NO reverse prompt would be inserted automatically. In order to have some thing prefixes your input, use `--in-prefix` as intended.  As an example, for Vicuna, previously you may have
```
-r 'USER:' --in-prefix ' '
```
after this PR, you only need
```
--in-prefix 'USER: '
```
which is because Vicuna is capable of generating EOS. In fact, the latter worked before this PR too. Same things for Wizarldlm or other model that uses EOS to signal the end of generation.

**Comment by JackJollimore** [comment]
> This is great! FYI as an additional benefit, this unblocks using `--grammar` in interactive mode as an alternative for several prompt options like `--input-prefix`, `--input-suffix` and `--reverse-prompt` (except for the unnecessary newline):
> 
> ```
> $ ./main -m $LLAMA2_13B_Q4_0  -i  \
>   --grammar 'root ::= "### RESPONSE: *" [a-z]+ (" " [a-z]+)* "* " [^\r\n]+ "\n### HUMAN: "' \


@ejones Thanks again for this. I'm trying to wrap my head around the basics that make --grammar work. Here's my attempt:
```
./main -m ~/Luna-AI-Llama2-Uncensored.ggmlv3.q4_0.bin -t 3 -b 7 -i -p "A chat between a curious user and an artificial intelligence assistant. The assistant gives helpful, detailed, and polite answers to the user's questions. USER: Hi, how are you? " --grammar 'root ::= "ASSISTANT: " [a-z]+ (" " [a-z]+)* [^\r\n]+ "\nUSER: "'
```
This functions, but I'm trying to understand what's happening. Here's my thought: root equals ASSISTANT who uses characters a to z including spaces and more a to z characters multiplied with (^?), a return keypress, a newline added to a second newline before User.

Will you help me understand what's actually happening inside `'root ::= "ASSISTANT: " [a-z]+ (" " [a-z]+)* [^\r\n]+ "\nUSER: "'`?


Thank you.

**Pull request merged**